transport-classes-epoll/src/main/java/io/netty5/channel/epoll/AbstractEpollStreamChannel.java

```diff
@@ -398,115 +393,37 @@ protected Object filterOutboundMessage(Object msg) {
-                shutdownInput0(promise);
-            } else {
-                loop.execute(() -> shutdownInput0(promise));
-            }
+    protected void doShutdown(ChannelShutdownDirection direction) throws Exception {
+        switch (direction) {
+            case Outbound:
+                socket.shutdown(false, true);
+                break;
+            case Inbound:
+                try {
+                    socket.shutdown(true, false);
+                } catch (NotYetConnectedException ignore) {
+                    // We attempted to shutdown and failed, which means the input has already effectively been
+                    // shutdown.
+                }
+                break;
+            default:
+                throw new IllegalStateException();
         }
-        return promise.asFuture();
     }
 
     @Override
-    public Future<Void> shutdown() {
-        return shutdown(newPromise());
-    }
-
-    @Override
-    public Future<Void> shutdown(Promise<Void> promise) {
-        Future<Void> shutdownOutputFuture = shutdownOutput();
-        if (shutdownOutputFuture.isDone()) {
-            shutdownOutputDone(promise, shutdownOutputFuture);
-        } else {
-            shutdownOutputFuture.addListener(promise, this::shutdownOutputDone);
-        }
-        return promise.asFuture();
-    }
-
-    private void shutdownOutputDone(Promise<Void> promise, Future<?> shutdownOutputFuture) {
-        Future<Void> shutdownInputFuture = shutdownInput();
-        if (shutdownInputFuture.isDone()) {
-            shutdownDone(shutdownOutputFuture, shutdownInputFuture, promise);
-        } else {
-            shutdownInputFuture.addListener(shutdownInputFuture1 ->
-                    shutdownDone(shutdownOutputFuture, shutdownInputFuture1, promise));
+    public boolean isShutdown(ChannelShutdownDirection direction) {
+        if (!isActive()) {
+            return true;
         }
-    }
-
-    private static void shutdownDone(Future<?> shutdownOutputFuture,
-                              Future<?> shutdownInputFuture,
-                              Promise<Void> promise) {
-        Throwable shutdownOutputCause = shutdownOutputFuture.cause();
-        Throwable shutdownInputCause = shutdownInputFuture.cause();
-        if (shutdownOutputCause != null) {
-            if (shutdownInputCause != null) {
-                logger.debug("Exception suppressed because a previous exception occurred.",
-                        shutdownInputCause);
-            }
-            promise.setFailure(shutdownOutputCause);
-        } else if (shutdownInputCause != null) {
-            promise.setFailure(shutdownInputCause);
-        } else {
-            promise.setSuccess(null);
+        switch (direction) {
+            case Outbound:
+                return socket.isOutputShutdown();
+            case Inbound:
+                return socket.isInputShutdown();
+            default:
+                throw new IllegalStateException();
```

Comment:
This should probably be an `IllegalArgumentException` instead.

Reply:
I am not sure about this... basically if this happens something is seriously wrong as we should never hit this default branch. Maybe `new AssertError()` ?

Reply:
That also works